### ISSUE TITLE
neo4j.bolt.connection.ServiceUnavailable

### ISSUE THREAD

**Issue by Tazeg** [neo4j/neo4j-python-driver#128]
Having this issue :
http://stackoverflow.com/questions/41765260/cant-connect-with-bolt-to-a-docker-neo4j

**Comment by Tazeg** [comment]
Post edited with more details.

**Comment by zebRAW** [comment]
I am having an issue that might be related:
http://stackoverflow.com/questions/42225439/neo4j-closes-session-unexpectedly-when-accessed-locally

EDIT: Using version 1.0.2 instead of 1.1.0 seems to have fixed the problem.

**Comment by yantisj** [comment]
I am having the same issue with 1.1.0 and 1.2.0rc1, the and receive this error with 1.2.0rc1: "neo4j.exceptions.ServiceUnavailable: Connection pool closed" immediately after establishing a connection. Everything works fine with 1.0.2 as others have stated, and I am running v3.1.1 on MacOS.

Inside the method where I establish my connection, I can make a query, but as soon as I move outside that scope, it breaks. I assume the connection pool is doing some sort of checking on whether to keep the connection established, and is closing the connection?

Basically this works:
def get_bolt_session():
    driver = GraphDatabase.driver(bolt_url, auth=auth_token)
    test = driver.session.run('MATCH (n {name:"hal"}) SET n.test = "test" RETURN n')
    return driver.session

But when I return the session to another method, it breaks:
bolt_session = get_bolt_session()
bolt_session.run('MATCH (n {name:"hal"}) SET n.test = "test" RETURN n')
neo4j.exceptions.ServiceUnavailable: Connection pool closed

**Comment by technige** [comment]
In your code, you are instantiating a `Driver` object which then falls out of scope when the function returns. The `Driver` holds the connection pool and should be maintained in global or class-wide scope for as long as you want to use sessions that it spawns. Closing the `Driver` (explicitly or implicitly) also closes the connection pool and any connections currently in use.

**Comment by yantisj** [comment]
Much appreciated, I was able to rework my code to use the driver instead of the session creating a session on demand, which makes sense now. Everything seems to be working.

**Comment by uludag** [comment]
I'm happy that I read @technige's comment. I was starting a neo4j session in a python class constructor and was only saving the session pointer not the driver pointer. I lost some hours with the ServiceUnavailable error which was not helping   me to see the actual reason of connection failure

**Comment by m-gerami** [comment]
Hi
I have a problem with bolt connection. I have unzipped neo4j in windows 10 and after installing services and starting neo4j, I opened the browser. Everything seems good, but when I write a Cypher query in the browser it shows an error: "Bolt connection Error, no connection found, did you connect to Neo4j?"
Would you please help me to solve this problem?

**Comment by openmindgap** [comment]
> Hi
> I have a problem with bolt connection. I have unzipped neo4j in windows 10 and after installing services and starting neo4j, I opened the browser. Everything seems good, but when I write a Cypher query in the browser it shows an error: "Bolt connection Error, no connection found, did you connect to Neo4j?"
> Would you please help me to solve this problem?

No response for over 2 years? I am getting the same thing in v4.0.3 and there is nothing online about it.
